File: transpiler/src/main/java/com/here/ivi/api/generator/cbridge/CBridgeModelBuilder.java
/*
 * Copyright (C) 2017 HERE Global B.V. and its affiliate(s). All rights reserved.
 *
 * This software, including documentation, is protected by copyright controlled by
 * HERE Global B.V. All rights are reserved. Copying, including reproducing, storing,
 * adapting or translating, any or all of this material requires the prior written
 * consent of HERE Global B.V. This material also contains confidential information,
 * which may not be disclosed to others without prior written consent of HERE Global B.V.
 *
 */

package com.here.ivi.api.generator.cbridge;

import static com.here.ivi.api.generator.cbridge.CppTypeInfo.TypeCategory.*;

import com.google.common.annotations.VisibleForTesting;
import com.here.ivi.api.common.CollectionsHelper;
import com.here.ivi.api.generator.common.modelbuilder.AbstractModelBuilder;
import com.here.ivi.api.generator.common.modelbuilder.ModelBuilderContextStack;
import com.here.ivi.api.generator.cpp.CppModelBuilder;
import com.here.ivi.api.generator.cpp.CppNameRules;
import com.here.ivi.api.generator.swift.SwiftModelBuilder;
import com.here.ivi.api.model.cbridge.*;
import com.here.ivi.api.model.cbridge.IncludeResolver.HeaderType;
import com.here.ivi.api.model.common.Include;
import com.here.ivi.api.model.cpp.CppField;
import com.here.ivi.api.model.cpp.CppMethod;
import com.here.ivi.api.model.franca.FrancaDeploymentModel;
import com.here.ivi.api.model.swift.SwiftField;
import com.here.ivi.api.model.swift.SwiftMethod;
import com.here.ivi.api.model.swift.SwiftProperty;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.franca.core.franca.*;

public class CBridgeModelBuilder extends AbstractModelBuilder<CElement> {

  private final FrancaDeploymentModel deploymentModel;
  private final IncludeResolver resolver;
  private final CppModelBuilder cppBuilder;
  private final SwiftModelBuilder swiftBuilder;
  public final Map<String, CArray> arraysCollector = new HashMap<>();

  public CBridgeModelBuilder(
      final FrancaDeploymentModel deploymentModel,
      IncludeResolver includeResolver,
      CppModelBuilder cppBuilder,
      SwiftModelBuilder swiftBuilder) {
    this(
        new ModelBuilderContextStack<>(),
        deploymentModel,
        includeResolver,
        cppBuilder,
        swiftBuilder);
  }

  @VisibleForTesting
  CBridgeModelBuilder(
      final ModelBuilderContextStack<CElement> contextStack,
      final FrancaDeploymentModel deploymentModel,
      final IncludeResolver includeResolver,
      final CppModelBuilder cppBuilder,
      final SwiftModelBuilder swiftBuilder) {
    super(contextStack);
    this.deploymentModel = deploymentModel;
    this.resolver = includeResolver;
    this.cppBuilder = cppBuilder;
    this.swiftBuilder = swiftBuilder;
  }

  @Override
  public void startBuilding(FInterface francaInterface) {
    super.startBuilding(francaInterface);
    storeResult(CTypeMapper.createCustomTypeInfo(resolver, francaInterface, CLASS));
  }

  @Override
  public void finishBuilding(FEnumerationType enumerationType) {
    storeResult(new CEnum(CBridgeNameRules.getEnumName(enumerationType)));
    super.finishBuilding(enumerationType);
  }

  @Override
  public void finishBuilding(FInterface francaInterface) {
    CppTypeInfo classInfo =
        CollectionsHelper.getFirstOfType(getCurrentContext().currentResults, CppTypeInfo.class);
    CInterface cInterface =
        finishBuildingInterfaceOrTypeCollection(
            CBridgeNameRules.getInterfaceName(francaInterface), classInfo, francaInterface);

    if (deploymentModel.isInterface(francaInterface)) {
      cInterface.functionTableName = CBridgeNameRules.getFunctionTableName(francaInterface);
      cInterface.implementationIncludes.add(
          Include.createInternalInclude(CBridgeComponents.PROXY_CACHE_FILENAME));
    }

    storeResult(cInterface);
    super.finishBuilding(francaInterface);
  }

  @Override
  public void finishBuilding(FTypeCollection francaTypeCollection) {
    CInterface cInterface =
        finishBuildingInterfaceOrTypeCollection(
            francaTypeCollection.getName(), null, francaTypeCollection);
    storeResult(cInterface);
    closeContext();
  }

  private CInterface finishBuildingInterfaceOrTypeCollection(
      final String name, final CppTypeInfo selfType, final FTypeCollection francaTypeCollection) {

    CInterface parentClass = getPreviousResult(CInterface.class);
    CInterface cInterface = new CInterface(name, selfType);
    if (parentClass != null) {
      cInterface.inheritedFunctions.addAll(parentClass.inheritedFunctions);
      cInterface.inheritedFunctions.addAll(parentClass.functions);
    }
    cInterface.functions.addAll(getPreviousResults(CFunction.class));
    cInterface.structs.addAll(getPreviousResults(CStruct.class));
    cInterface.enumerators.addAll(getPreviousResults(CEnum.class));

    cInterface.headerIncludes.addAll(CBridgeComponents.collectHeaderIncludes(cInterface));
    cInterface.implementationIncludes.addAll(
        CBridgeComponents.collectImplementationIncludes(cInterface));

    cInterface.implementationIncludes.add(
        resolver.resolveInclude(francaTypeCollection, HeaderType.CBRIDGE_PUBLIC_HEADER));
    cInterface.privateHeaderIncludes.addAll(
        CBridgeComponents.collectPrivateHeaderIncludes(cInterface));
    return cInterface;
  }

  @Override
  public void finishBuilding(FMethod francaMethod) {

    SwiftMethod swiftMethod = swiftBuilder.getFinalResult(SwiftMethod.class);
    CppMethod cppMethod = cppBuilder.getFinalResult(CppMethod.class);
    List<CInParameter> inParams = getPreviousResults(CInParameter.class);
    COutParameter returnParam =
        CollectionsHelper.getFirstOfType(
            getCurrentContext().previousResults, COutParameter.class, new COutParameter());

    CFunction.CFunctionBuilder methodBuilder =
        CFunction.builder(swiftMethod.cShortName)
            .nestedSpecifier(swiftMethod.cNestedSpecifier)
            .delegateCall(cppMethod.fullyQualifiedName)
            .parameters(inParams)
            .returnType(returnParam.mappedType)
            .error(
                francaMethod.getErrorEnum() != null
                    ? CTypeMapper.createErrorTypeInfo(resolver, francaMethod.getErrorEnum())
                    : null)
            .delegateCallIncludes(
                Collections.singleton(
                    resolver.resolveInclude(francaMethod, HeaderType.BASE_API_HEADER)))
            .functionName(cppMethod.name);

    if (!deploymentModel.isStatic(francaMethod)) {
      CppTypeInfo classInfo =
          CollectionsHelper.getFirstOfType(getParentContext().currentResults, CppTypeInfo.class);
      CInParameter parameterSelf = new CInParameter("_instance", classInfo);
      methodBuilder
          .selfParameter(parameterSelf)
          .functionName(CppNameRules.getMethodName(francaMethod.getName()));
    }

    storeResult(methodBuilder.build());
    closeContext();
  }

  @Override
  public void finishBuildingInputArgument(FArgument francaArgument) {
    storeResult(new CInParameter(francaArgument.getName(), getPreviousResult(CppTypeInfo.class)));
    closeContext();
  }

  @Override
  public void finishBuildingOutputArgument(FArgument francaArgument) {
    storeResult(new COutParameter("result", getPreviousResult(CppTypeInfo.class)));
    closeContext();
  }

  @Override
  public void finishBuilding(FStructType francaStruct) {

    CStruct cStruct =
        new CStruct(
            CBridgeNameRules.getStructBaseName(francaStruct),
            CBridgeNameRules.getBaseApiStructName(francaStruct),
            CTypeMapper.createCustomTypeInfo(resolver, francaStruct, STRUCT));

    CStruct parentStruct = getPreviousResult(CStruct.class);
    if (parentStruct != null) {
      cStruct.fields.addAll(parentStruct.fields);
    }
    cStruct.fields.addAll(getPreviousResults(CField.class));

    storeResult(cStruct);
    super.finishBuilding(francaStruct);
  }

  @Override
  public void finishBuilding(FArrayType francaArray) {

    CppTypeInfo innerType = CTypeMapper.mapType(resolver, francaArray.getElementType());
    CArray cArray = CArrayMapper.createArrayDefinition(francaArray, innerType);
    arraysCollector.put(cArray.name, cArray);

    closeContext();
  }

  @Override
  public void finishBuilding(FField francaField) {
    CppField cppField = cppBuilder.getFinalResult(CppField.class);
    SwiftField swiftField = swiftBuilder.getFinalResult(SwiftField.class);
    storeResult(new CField(swiftField.name, cppField.name, getPreviousResult(CppTypeInfo.class)));
    super.finishBuilding(francaField);
  }

  @Override
  public void finishBuilding(FTypeRef typeRef) {
    CppTypeInfo type = CTypeMapper.mapType(resolver, typeRef);
    if (type.typeCategory == ARRAY) {
      CArray cArray =
          CArrayMapper.createArrayDefinition(typeRef.getDerived(), type.innerType, type);
      arraysCollector.put(cArray.name, cArray);
    }

    storeResult(type);
    super.finishBuilding(typeRef);
  }

  @Override
  public void finishBuilding(FAttribute attribute) {

    List<CppMethod> cppMethods =
        CollectionsHelper.getAllOfType(cppBuilder.getFinalResults(), CppMethod.class);

    SwiftProperty property = swiftBuilder.getFinalResult(SwiftProperty.class);

    CppTypeInfo classInfo =
        CollectionsHelper.getFirstOfType(getParentContext().currentResults, CppTypeInfo.class);
    CInParameter selfParameter = new CInParameter("_instance", classInfo);
    CppTypeInfo attributeTypeInfo = getPreviousResult(CppTypeInfo.class);

    SwiftMethod getterSwiftMethod = property.propertyAccessors.get(0);
    CFunction.CFunctionBuilder getterBuilder =
        CFunction.builder(getterSwiftMethod.cShortName)
            .nestedSpecifier(getterSwiftMethod.cNestedSpecifier)
            .returnType(attributeTypeInfo)
            .selfParameter(selfParameter)
            .functionName(cppMethods.get(0).name);
    storeResult(getterBuilder.build());

    if (!attribute.isReadonly()) {
      SwiftMethod setterSwiftMethod = property.propertyAccessors.get(1);
      CFunction.CFunctionBuilder setterBuilder =
          CFunction.builder(setterSwiftMethod.cShortName)
              .nestedSpecifier(setterSwiftMethod.cNestedSpecifier)
              .parameters(
                  Collections.singletonList(new CInParameter("newValue", attributeTypeInfo)))
              .selfParameter(selfParameter)
              .functionName(cppMethods.get(1).name);
      storeResult(setterBuilder.build());
    }

    super.finishBuilding(attribute);
  }
}

File: gluecodium/src/main/java/com/here/gluecodium/generator/cbridge/CBridgeModelBuilder.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.gluecodium.generator.cbridge

import com.here.gluecodium.common.ModelBuilderContextStack
import com.here.gluecodium.generator.common.modelbuilder.AbstractLimeBasedModelBuilder
import com.here.gluecodium.generator.cpp.CppIncludeResolver
import com.here.gluecodium.generator.cpp.CppModelBuilder
import com.here.gluecodium.generator.swift.SwiftModelBuilder
import com.here.gluecodium.model.cbridge.CElement
import com.here.gluecodium.model.cbridge.CEnum
import com.here.gluecodium.model.cbridge.CField
import com.here.gluecodium.model.cbridge.CFunction
import com.here.gluecodium.model.cbridge.CInterface
import com.here.gluecodium.model.cbridge.CParameter
import com.here.gluecodium.model.cbridge.CStruct
import com.here.gluecodium.model.cbridge.CType
import com.here.gluecodium.model.common.Include
import com.here.gluecodium.model.cpp.CppField
import com.here.gluecodium.model.cpp.CppFunctionTypeRef
import com.here.gluecodium.model.cpp.CppMethod
import com.here.gluecodium.model.cpp.CppParameter
import com.here.gluecodium.model.cpp.CppStruct
import com.here.gluecodium.model.cpp.CppTypeRef
import com.here.gluecodium.model.cpp.CppUsing
import com.here.gluecodium.model.lime.LimeAttributeType
import com.here.gluecodium.model.lime.LimeAttributeType.CPP
import com.here.gluecodium.model.lime.LimeAttributeType.SWIFT
import com.here.gluecodium.model.lime.LimeAttributeValueType
import com.here.gluecodium.model.lime.LimeAttributeValueType.SKIP
import com.here.gluecodium.model.lime.LimeContainerWithInheritance
import com.here.gluecodium.model.lime.LimeElement
import com.here.gluecodium.model.lime.LimeEnumeration
import com.here.gluecodium.model.lime.LimeField
import com.here.gluecodium.model.lime.LimeFunction
import com.here.gluecodium.model.lime.LimeInterface
import com.here.gluecodium.model.lime.LimeLambda
import com.here.gluecodium.model.lime.LimeNamedElement
import com.here.gluecodium.model.lime.LimeParameter
import com.here.gluecodium.model.lime.LimeProperty
import com.here.gluecodium.model.lime.LimeStruct
import com.here.gluecodium.model.lime.LimeTypeRef
import com.here.gluecodium.model.lime.LimeTypesCollection
import com.here.gluecodium.model.swift.SwiftField
import com.here.gluecodium.model.swift.SwiftMethod
import com.here.gluecodium.model.swift.SwiftParameter
import com.here.gluecodium.model.swift.SwiftProperty

class CBridgeModelBuilder(
    contextStack: ModelBuilderContextStack<CElement> = ModelBuilderContextStack(),
    private val limeReferenceMap: Map<String, LimeElement>,
    private val cppIncludeResolver: CppIncludeResolver,
    private val cppBuilder: CppModelBuilder,
    private val swiftBuilder: SwiftModelBuilder,
    private val typeMapper: CBridgeTypeMapper,
    private val internalNamespace: List<String>,
    private val buildTransientModel: (LimeNamedElement) -> List<CInterface>
) : AbstractLimeBasedModelBuilder<CElement>(contextStack) {

    override fun startBuilding(limeContainer: LimeContainerWithInheritance) {
        openContext()
        val cppTypeInfo = typeMapper.createCustomTypeInfo(limeContainer, isClass = true)
        storeResult(cppTypeInfo)
    }

    override fun startBuilding(limeStruct: LimeStruct) {
        openContext()
        val cppTypeInfo = typeMapper.createCustomTypeInfo(limeStruct)
        storeResult(cppTypeInfo)
    }

    override fun finishBuilding(limeContainer: LimeContainerWithInheritance) {
        val parentClass = limeContainer.parent?.type?.let { buildTransientModel(it).first() }
        val inheritedFunctions = mutableListOf<CFunction>()
        if (parentClass != null) {
            inheritedFunctions.addAll(parentClass.inheritedFunctions)
            inheritedFunctions.addAll(parentClass.functions)
        }
        val functionTableName =
            (limeContainer as? LimeInterface)?.let { CBridgeNameRules.getFunctionTableName(it) }

        val cInterface = CInterface(
            name = CBridgeNameRules.getInterfaceName(limeContainer),
            selfType = currentContext.currentResults.filterIsInstance<CppTypeInfo>().firstOrNull(),
            internalNamespace = internalNamespace,
            structs = getPreviousResults(CStruct::class.java),
            enums = getPreviousResults(CEnum::class.java),
            functions = getPreviousResults(CFunction::class.java),
            interfaces = getPreviousResults(CInterface::class.java),
            inheritedFunctions = inheritedFunctions,
            functionTableName = functionTableName,
            isEquatable = limeContainer.attributes.have(LimeAttributeType.EQUATABLE),
            isPointerEquatable = limeContainer.attributes.have(LimeAttributeType.POINTER_EQUATABLE),
            hasTypeRepository = limeContainer is LimeInterface || parentClass != null ||
                    limeContainer.visibility.isOpen
        )

        if (limeContainer is LimeInterface) {
            cInterface.implementationIncludes.add(
                Include.createInternalInclude(CBridgeComponents.PROXY_CACHE_FILENAME)
            )
        }
        cInterface.implementationIncludes.add(cppIncludeResolver.typeRepositoryInclude)
        cInterface.implementationIncludes.add(Include.createInternalInclude(CBridgeNameRules.TYPE_INIT_REPOSITORY))
        cInterface.implementationIncludes.add(Include.createInternalInclude(CBridgeComponents.WRAPPER_CACHE_HEADER))

        storeResult(cInterface)
        closeContext()
    }

    override fun finishBuilding(limeTypes: LimeTypesCollection) {
        val cInterface = CInterface(
            name = CBridgeNameRules.getInterfaceName(limeTypes),
            internalNamespace = internalNamespace,
            structs = getPreviousResults(CStruct::class.java),
            enums = getPreviousResults(CEnum::class.java)
        )

        storeResult(cInterface)
        closeContext()
    }

    override fun finishBuilding(limeMethod: LimeFunction) {
        var parameterSelf: CParameter? = null
        if (!limeMethod.isStatic) {
            val cppTypeInfo = parentContext!!.currentResults.filterIsInstance<CppTypeInfo>().first()
            parameterSelf = CParameter("_instance", cppTypeInfo)
        }

        val swiftMethod = swiftBuilder.getFinalResult(SwiftMethod::class.java)
        val cppMethod = cppBuilder.getFinalResult(CppMethod::class.java)
        val limeParent =
            limeReferenceMap[limeMethod.path.parent.toString()] as LimeNamedElement
        val returnType = when {
            limeParent !is LimeStruct && limeMethod.isConstructor ->
                typeMapper.createCustomTypeInfo(limeParent, isClass = true)
            else -> typeMapper.mapType(limeMethod.returnType.typeRef, cppMethod.returnType)
        }

        val limeErrorTypeRef = limeMethod.exception?.errorType
        val cppErrorType = cppMethod.errorType
        val errorType = if (limeErrorTypeRef != null && cppErrorType != null) {
            typeMapper.mapType(limeErrorTypeRef, cppErrorType)
        } else null

        val errorTypeIsEnum = limeErrorTypeRef?.type?.actualType is LimeEnumeration
        val result = CFunction(
            shortName = swiftMethod.cShortName,
            nestedSpecifier = swiftMethod.cNestedSpecifier,
            returnType = returnType,
            parameters = getPreviousResults(CParameter::class.java),
            selfParameter = parameterSelf,
            delegateCall = cppMethod.fullyQualifiedName,
            delegateCallIncludes = cppIncludeResolver.resolveIncludes(limeMethod).toSet(),
            functionName = cppMethod.name,
            cppReturnTypeName = cppMethod.returnType.fullyQualifiedName,
            isConst = limeMethod.attributes.have(CPP, LimeAttributeValueType.CONST),
            error = errorType,
            errorTypeIsEnum = errorTypeIsEnum,
            isSkipped = limeMethod.attributes.have(SWIFT, SKIP)
        )

        storeResult(result)
        closeContext()
    }

    override fun finishBuilding(limeParameter: LimeParameter) {
        var cppTypeInfo = getPreviousResult(CppTypeInfo::class.java)
        if (limeParameter.typeRef.isNullable) {
            val cppParameter = cppBuilder.getFinalResult(CppParameter::class.java)
            cppTypeInfo = typeMapper.createNullableTypeInfo(cppTypeInfo, cppParameter.type)
        }
        val swiftParameter = swiftBuilder.getFinalResult(SwiftParameter::class.java)
        val result = CParameter(swiftParameter.name, cppTypeInfo)

        storeResult(result)
        closeContext()
    }

    override fun finishBuilding(limeStruct: LimeStruct) {
        val cppStruct = cppBuilder.getFinalResult(CppStruct::class.java)
        val cStruct = CStruct(
            name = CBridgeNameRules.getTypeName(limeStruct),
            baseApiName = cppStruct.fullyQualifiedName,
            mappedType = currentContext.currentResults.filterIsInstance<CppTypeInfo>().first(),
            hasImmutableFields = cppStruct.hasImmutableFields,
            fields = getPreviousResults(CField::class.java),
            methods = getPreviousResults(CFunction::class.java),
            structs = getPreviousResults(CStruct::class.java),
            interfaces = getPreviousResults(CInterface::class.java),
            enums = getPreviousResults(CEnum::class.java)
        )

        storeResult(cStruct)
        closeContext()
    }

    override fun finishBuilding(limeField: LimeField) {
        val cppField = cppBuilder.getFinalResult(CppField::class.java)
        val swiftField = swiftBuilder.getFinalResult(SwiftField::class.java)

        var cppTypeInfo = getPreviousResult(CppTypeInfo::class.java)
        if (limeField.typeRef.isNullable) {
            cppTypeInfo = typeMapper.createNullableTypeInfo(cppTypeInfo, cppField.type)
        }

        val cField = CField(
            swiftLayerName = swiftField.name,
            baseLayerName = cppField.name,
            type = cppTypeInfo,
            baseLayerGetterName = cppField.getterName,
            baseLayerSetterName = cppField.setterName
        )

        storeResult(cField)
        closeContext()
    }

    override fun finishBuilding(limeEnumeration: LimeEnumeration) {
        val result = CEnum(
            CBridgeNameRules.getTypeName(limeEnumeration),
            typeMapper.createEnumTypeInfo(limeEnumeration)
        )

        storeResult(result)
        closeContext()
    }

    override fun finishBuilding(limeProperty: LimeProperty) {
        val cppMethods = cppBuilder.finalResults.filterIsInstance<CppMethod>()
        val swiftProperty = swiftBuilder.getFinalResult(SwiftProperty::class.java)

        val classInfo = parentContext!!.currentResults.filterIsInstance<CppTypeInfo>().first()
        val selfParameter = when {
            limeProperty.isStatic -> null
            else -> CParameter("_instance", classInfo)
        }

        val cppGetterMethod = cppMethods.first()
        var attributeTypeInfo = getPreviousResult(CppTypeInfo::class.java)
        if (limeProperty.typeRef.isNullable) {
            attributeTypeInfo = typeMapper.createNullableTypeInfo(
                attributeTypeInfo,
                cppGetterMethod.returnType
            )
        }

        val getterSwiftMethod = swiftProperty.getter
        val getterFunction = CFunction(
            shortName = getterSwiftMethod.cShortName,
            nestedSpecifier = getterSwiftMethod.cNestedSpecifier,
            returnType = attributeTypeInfo,
            parameters = emptyList(),
            selfParameter = selfParameter,
            delegateCall = cppGetterMethod.fullyQualifiedName,
            delegateCallIncludes = cppIncludeResolver.resolveIncludes(limeProperty).toSet(),
            functionName = cppGetterMethod.name,
            cppReturnTypeName = cppGetterMethod.returnType.fullyQualifiedName,
            isConst = true,
            isSkipped = limeProperty.attributes.have(SWIFT, SKIP)
        )
        storeResult(getterFunction)

        if (limeProperty.setter != null) {
            val setterSwiftMethod = swiftProperty.setter
            val cppSetterMethod = cppMethods[1]
            val setterFunction = CFunction(
                shortName = setterSwiftMethod?.cShortName,
                nestedSpecifier = setterSwiftMethod?.cNestedSpecifier,
                returnType = CppTypeInfo(CType.VOID),
                parameters = listOf(CParameter("newValue", attributeTypeInfo)),
                selfParameter = selfParameter,
                delegateCall = cppSetterMethod.fullyQualifiedName,
                delegateCallIncludes = cppIncludeResolver.resolveIncludes(limeProperty).toSet(),
                functionName = cppSetterMethod.name,
                cppReturnTypeName = cppSetterMethod.returnType.fullyQualifiedName,
                isSkipped = limeProperty.attributes.have(SWIFT, SKIP)
            )
            storeResult(setterFunction)
        }

        closeContext()
    }

    override fun finishBuilding(limeLambda: LimeLambda) {
        val cppUsing = cppBuilder.getFinalResult(CppUsing::class.java)
        val cppFunction = cppUsing.definition as CppFunctionTypeRef
        val parameters = limeLambda.parameters.mapIndexed { index, parameter ->
            CParameter(
                "p$index",
                typeMapper.mapType(parameter.typeRef, cppFunction.parameters[index])
            )
        }
        val selfType = CBridgeTypeMapper.createFunctionalTypeInfo(cppUsing)
        val cFunction = CFunction(
            shortName = "call",
            nestedSpecifier = CBridgeNameRules.getNestedSpecifierString(limeLambda),
            returnType = typeMapper.mapType(limeLambda.returnType.typeRef, cppFunction.returnType),
            parameters = parameters,
            selfParameter = CParameter("_instance", selfType),
            delegateCallIncludes = cppIncludeResolver.resolveIncludes(limeLambda).toSet(),
            functionName = "operator()",
            cppReturnTypeName = cppFunction.returnType.fullyQualifiedName
        )

        val cElement = CInterface(
            name = CBridgeNameRules.getInterfaceName(limeLambda),
            selfType = selfType,
            functionTableName = CBridgeNameRules.getFunctionTableName(limeLambda),
            internalNamespace = internalNamespace,
            functions = listOf(cFunction),
            isFunctionalInterface = true
        )
        cElement.implementationIncludes +=
            Include.createInternalInclude(CBridgeComponents.PROXY_CACHE_FILENAME)

        storeResult(cElement)
        closeContext()
    }

    override fun finishBuilding(limeTypeRef: LimeTypeRef) {
        val cppType = cppBuilder.getFinalResult(CppTypeRef::class.java)
        val cppTypeInfo = typeMapper.mapType(limeTypeRef, cppType)

        storeResult(cppTypeInfo)
        closeContext()
    }
}
